### PR TITLE
qt: Cleanup MacDockIconHandler class

### PR DESCRIPTION
**Pull request by hebasto**
The Objective-C `MacDockIconHandler` class currently does such jobs on macOS:

1. Set the Dock icon
2. Support the Dock icon menu
3. Handle the Dock icon click event

Qt does the first job natively.
Qt has native support the Dock icon menu only since version 5.2
Qt does not handle `applicationShouldHandleReopen` event. And this brakes all Qt support for the Dock icon click event.

This PR:
- removes Objective-C code for the Dock icon setting. Qt `setWindowIcon()` does this work
- removes Objective-C code for the Dock icon menu. Qt `setAsDockMenu()` does this work
- uses Qt signal for the Dock icon click event

### CODE REVIEW

**Comment by promag** [comment]
Concept ACK.

**Comment by practicalswift** [comment]
Concept ACK

Net change: -189 lines. Nice cleanup!

**Comment by fanquake** [comment]
Concept ACK

There's at least one bug here. Using e0d1883, if you close the main window using the red cross, you can no longer bring the window back by clicking on the dock icon. 
This might be solved by one of the other GUI related PRs, possibly #14123.

**Comment by ismail120572** [comment]
Who do that

Ismail

Pada tanggal Sel, 30 Okt 2018 18:02, fanquake <notifications@github.com>
menulis:

**Comment by ismail120572** [comment]
Ismail

Pada tanggal Sel, 30 Okt 2018 18:02, fanquake <notifications@github.com>
menulis:

**Comment by promag** [comment]
> This might be solved by one of the other GUI related PRs, possibly #14123.

@fanquake no it doesn't solve it. But using a dock menu action, like "Send", it reopens.

**Comment by hebasto** [comment]
@fanquake @promag @practicalswift 
Thank you for your reviews.
The original goal of this PR (to completely remove `MacDockIconHandler` class) has revealed inaccessible.
So PR has be reworked. Would you mind re-reviewing?

File: src/qt/macdockiconhandler.mm [hunk 1]
```diff
@@ -49,61 +44,22 @@ void setupDockClickHandler() {
     setupDockClickHandler();
     this->m_dummyWidget = new QWidget();
     this->m_dockMenu = new QMenu(this->m_dummyWidget);
-    this->setMainWindow(nullptr);
 #if QT_VERSION >= 0x050200
```
Comment (fanquake) [inline src/qt/macdockiconhandler.mm]: You can drop the `QT_VERSION` check here

**Comment by hebasto** [comment]
@fanquake Your comment is addressed. Please re-review.

File: src/qt/macdockiconhandler.mm [hunk 2]
```diff
@@ -34,74 +28,13 @@ void setupDockClickHandler() {
         id delegate = objc_msgSend(appInst, sel_registerName("delegate"));
         Class delClass = (Class)objc_msgSend(delegate,  sel_registerName("class"));
         SEL shouldHandle = sel_registerName("applicationShouldHandleReopen:hasVisibleWindows:");
-        if (class_getInstanceMethod(delClass, shouldHandle))
-            class_replaceMethod(delClass, shouldHandle, (IMP)dockClickHandler, "B@:");
-        else
-            class_addMethod(delClass, shouldHandle, (IMP)dockClickHandler,"B@:");
+        class_replaceMethod(delClass, shouldHandle, (IMP)dockClickHandler, "B@:");
```
Comment (Sjors) [inline src/qt/macdockiconhandler.mm]: Why did/do we need this `class_replaceMethod` stuff?

**Review by Sjors — APPROVED** [review]
tACK efb681d on macOS 10.14.1, thanks for reducing the amount of OS specific code we need to reason about!

File: src/qt/macdockiconhandler.mm (hunk 2 above)
Comment (hebasto) [reply src/qt/macdockiconhandler.mm]: This implements the method of the application delegate which handles the `reopen application` event (i.e., the user clicks application icon in the Dock).
Ref: [How Cocoa Applications Handle Apple Events](https://developer.apple.com/library/archive/documentation/Cocoa/Conceptual/ScriptableCocoaApplications/SApps_handle_AEs/SAppsHandleAEs.html)

**Comment by hebasto** [comment]
Fixed bug with hided/inactive/obscured main window: `toggleHidden()` does not work on macOS properly.

**Comment by jonasschnelli** [comment]
This needs proper testing on macOS 10.9 up till 10.14. Also, I'm not sure that this works reliable with older Qt5 versions. We should probably also test against Qt5.4 till Qt5.9.

**Comment by hebasto** [comment]
@jonasschnelli 
> This needs proper testing on macOS 10.9 up till 10.14. Also, I'm not sure that this works reliable with older Qt5 versions. We should probably also test against Qt5.4 till Qt5.9.


https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes.md:
> From 0.17.0 onwards macOS <10.10 is no longer supported. 0.17.0 is built using Qt 5.9.x, which doesn't support versions of macOS older than 10.10.

What is the reason to choose so wide range of the tested platforms?

**Comment by ismail120572** [comment]
Dikirim dari ponsel cerdas Samsung Galaxy saya.
-------- Pesan asli --------Dari: Jonas Schnelli <notifications@github.com> Tanggal: 04/11/18  21:03  (GMT+07:00) Ke: bitcoin/bitcoin <bitcoin@noreply.github.com> Cc: Ismail <nusagala@gmail.com>, Comment <comment@noreply.github.com> Subjek: Re: [bitcoin/bitcoin] qt: Cleanup MacDockIconHandler class (#14597) 
This needs proper testing on macOS 10.9 up till 10.14. Also, I'm not sure that this works reliable with older Qt5 versions. We should probably also test against Qt5.4 till Qt5.9.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub, or mute the thread.
{"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/bitcoin/bitcoin","title":"bitcoin/bitcoin","subtitle":"GitHub repository","main_image_url":"https://assets-cdn.github.com/images/email/message_cards/header.png","avatar_image_url":"https://assets-cdn.github.com/images/email/message_cards/avatar.png","action":{"name":"Open in GitHub","url":"https://github.com/bitcoin/bitcoin"}},"updates":{"snippets":[{"icon":"PERSON","message":"@jonasschnelli in #14597: This needs proper testing on macOS 10.9 up till 10.14. Also, I'm not sure that this works reliable with older Qt5 versions. We should probably also test against Qt5.4 till Qt5.9."}],"action":{"name":"View Pull Request","url":"https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-435672243"}}}
[
{
"@context": "http://schema.org",
"@type": "EmailMessage",
"potentialAction": {
"@type": "ViewAction",
"target": "https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-435672243",
"url": "https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-435672243",
"name": "View Pull Request"
},
"description": "View this Pull Request on GitHub",
"publisher": {
"@type": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
},
{
"@type": "MessageCard",
"@context": "http://schema.org/extensions",
"hideOriginalBody": "false",
"originator": "AF6C5A86-E920-430C-9C59-A73278B5EFEB",
"title": "Re: [bitcoin/bitcoin] qt: Cleanup MacDockIconHandler class (#14597)",
"sections": [
{
"text": "",
"activityTitle": "**Jonas Schnelli**",
"activityImage": "https://assets-cdn.github.com/images/email/message_cards/avatar.png",
"activitySubtitle": "@jonasschnelli",
"facts": [

]
}
],
"potentialAction": [
{
"name": "Add a comment",
"@type": "ActionCard",
"inputs": [
{
"isMultiLine": true,
"@type": "TextInput",
"id": "IssueComment",
"isRequired": false
}
],
"actions": [
{
"name": "Comment",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"IssueComment\",\n\"repositoryFullName\": \"bitcoin/bitcoin\",\n\"issueId\": 14597,\n\"IssueComment\": \"{{IssueComment.value}}\"\n}"
}
]
},
{
"name": "Close pull request",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"PullRequestClose\",\n\"repositoryFullName\": \"bitcoin/bitcoin\",\n\"pullRequestId\": 14597\n}"
},
{
"targets": [
{
"os": "default",
"uri": "https://github.com/bitcoin/bitcoin/pull/14597#issuecomment-435672243"
}
],
"@type": "OpenUri",
"name": "View on GitHub"
},
{
"name": "Unsubscribe",
"@type": "HttpPOST",
"target": "https://api.github.com",
"body": "{\n\"commandName\": \"MuteNotification\",\n\"threadId\": 402189341\n}"
}
],
"themeColor": "26292E"
}
]

**Comment by ismail120572** [comment]
tidak sma sekali tidak keberatan justru saya senang ada yang meperhatian
seperti anda hingga terbuka transapasi
sekali lagi terima kasisih

Pada tanggal Kam, 1 Nov 2018 pukul 04.33 Hennadii Stepanov <
notifications@github.com> menulis:

**Comment by jonasschnelli** [comment]
> > This needs proper testing on macOS 10.9 up till 10.14. Also, I'm not sure that this works reliable with older Qt5 versions. We should probably also test against Qt5.4 till Qt5.9.
> https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes.md:

The "official" binaries are using Qt5.9 which doesn't mean that we drop support self-compilation with an older Qt version.

We should at least do some tests on something between OSX 10.10 and 10.14 with Qt5.5 and Qt5.9 (and not just only 10.14&Qt5.9).

**Pull request merged**